Non-interactive use leads to executable missing the project as a build dependency

I want to run cabal non-interactively to scaffold my project. It should include a library, executable, and test suite. I compiled cff9b1a7c6188796873ffcb7ba9332dca718548f (i.e. all v2 commands), run cabal init, and cabal run and I get an unfortunate error due to a missing `build-depend` in the executable stanza.

**To Reproduce**

```
mkdir sample-project
cd sample-project
cabal init --libandexe --test -n
cabal run sample-project
```

Running `cabal run sample-project` I get:

```
app/Main.hs:3:1: error:
    Could not load module ‘MyLib’
    It is a member of the hidden package ‘sample-project-0.1.0.0’.
    Perhaps you need to add ‘sample-project’ to the build-depends in your .cabal file.
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
  |
3 | import qualified MyLib (someFunc)
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Expected behavior**

`cabal run sample-project` should print

```
Hello, Haskell!
someFunc
```

**System information**
 - Pop!_OS 21.04
 - cabal 3.7 latest master commit cff9b1a7c6188796873ffcb7ba9332dca718548f and GHC 8.8.4

--- a/cabal-install/src/Distribution/Client/Init/Interactive/Command.hs
+++ b/cabal-install/src/Distribution/Client/Init/Interactive/Command.hs
@@ -149,19 +149,6 @@ createProject v pkgIx srcDb initFlags = do
       return $ ProjectSettings
         (mkOpts comments cabalSpec) pkgDesc
         Nothing Nothing testTarget
-  where
-    -- Add package name as dependency of test suite
-    --
-    addLibDepToTest _ Nothing = Nothing
-    addLibDepToTest n (Just t) = Just $ t
-      { _testDependencies = _testDependencies t ++ [mkPackageNameDep n]
-      }
-
-    -- Add package name as dependency of executable
-    --
-    addLibDepToExe n exe = exe
-      { _exeDependencies = _exeDependencies exe ++ [mkPackageNameDep n]
-      }
 
 -- -------------------------------------------------------------------- --
 -- Target and pkg description generation

--- a/cabal-install/src/Distribution/Client/Init/NonInteractive/Command.hs
+++ b/cabal-install/src/Distribution/Client/Init/NonInteractive/Command.hs
@@ -115,7 +115,8 @@ createProject comp v pkgIx srcDb initFlags = do
   case pkgType of
     Library -> do
       libTarget <- genLibTarget initFlags comp pkgIx cabalSpec
-      testTarget <- genTestTarget initFlags comp pkgIx cabalSpec
+      testTarget <- addLibDepToTest pkgName <$>
+        genTestTarget initFlags comp pkgIx cabalSpec
 
       return $ ProjectSettings
         (mkOpts comments cabalSpec) pkgDesc
@@ -130,8 +131,10 @@ createProject comp v pkgIx srcDb initFlags = do
 
     LibraryAndExecutable -> do
       libTarget <- genLibTarget initFlags comp pkgIx cabalSpec
-      exeTarget <- genExeTarget initFlags comp pkgIx cabalSpec
-      testTarget <- genTestTarget initFlags comp pkgIx cabalSpec
+      exeTarget <- addLibDepToExe pkgName <$>
+        genExeTarget initFlags comp pkgIx cabalSpec
+      testTarget <- addLibDepToTest pkgName <$>
+        genTestTarget initFlags comp pkgIx cabalSpec
 
       return $ ProjectSettings
         (mkOpts comments cabalSpec) pkgDesc (Just libTarget)

--- a/cabal-install/src/Distribution/Client/Init/Utils.hs
+++ b/cabal-install/src/Distribution/Client/Init/Utils.hs
@@ -18,6 +18,8 @@ module Distribution.Client.Init.Utils
 , fixupDocFiles
 , mkStringyDep
 , getBaseDep
+, addLibDepToExe
+, addLibDepToTest
 ) where
 
 
@@ -310,3 +312,18 @@ mkStringyDep = mkPackageNameDep . mkPackageName
 getBaseDep :: Interactive m => InstalledPackageIndex -> InitFlags -> m [Dependency]
 getBaseDep pkgIx flags = retrieveDependencies silent flags
   [(fromString "Prelude", fromString "Prelude")] pkgIx
+
+-- Add package name as dependency of test suite
+--
+addLibDepToTest :: PackageName -> Maybe TestTarget -> Maybe TestTarget
+addLibDepToTest _ Nothing = Nothing
+addLibDepToTest n (Just t) = Just $ t
+  { _testDependencies = _testDependencies t ++ [mkPackageNameDep n]
+  }
+
+-- Add package name as dependency of executable
+--
+addLibDepToExe :: PackageName -> ExeTarget -> ExeTarget
+addLibDepToExe n exe = exe
+  { _exeDependencies = _exeDependencies exe ++ [mkPackageNameDep n]
+  }

--- a/cabal-install/tests/UnitTests/Distribution/Client/Init/NonInteractive.hs
+++ b/cabal-install/tests/UnitTests/Distribution/Client/Init/NonInteractive.hs
@@ -22,7 +22,7 @@ import Distribution.ModuleName (fromString)
 import Distribution.Simple.Flag
 import Data.List (foldl')
 import qualified Data.Set as Set
-import Distribution.Client.Init.Utils (mkStringyDep)
+import Distribution.Client.Init.Utils (mkPackageNameDep, mkStringyDep)
 
 tests
     :: Verbosity
@@ -105,7 +105,7 @@ driverFunctionTest pkgIx srcDb comp = testGroup "createProject"
             _exeLanguage        exe @?= Haskell98
             _exeOtherModules    exe @?= []
             _exeOtherExts       exe @?= []
-            _exeDependencies    exe @?= []
+            _exeDependencies    exe @?! []
             _exeBuildTools      exe @?= []
 
             _testMainIs       test @?= HsFilePath "quxTest/Main.hs" Standard
@@ -113,8 +113,13 @@ driverFunctionTest pkgIx srcDb comp = testGroup "createProject"
             _testLanguage     test @?= Haskell98
             _testOtherModules test @?= []
             _testOtherExts    test @?= []
-            _testDependencies test @?= []
+            _testDependencies test @?! []
             _testBuildTools   test @?= []
+
+            assertBool "The library should be a dependency of the executable" $
+              mkPackageNameDep (_optPkgName opts) `elem` _exeDependencies exe
+            assertBool "The library should be a dependency of the test executable" $
+              mkPackageNameDep (_optPkgName opts) `elem` _testDependencies test
 
           Right (ProjectSettings _ _ lib exe test, _) -> do
             lib  @?! Nothing
@@ -185,7 +190,7 @@ driverFunctionTest pkgIx srcDb comp = testGroup "createProject"
             _exeLanguage        exe @?= Haskell98
             _exeOtherModules    exe @?= []
             _exeOtherExts       exe @?= []
-            _exeDependencies    exe @?= []
+            _exeDependencies    exe @?! []
             _exeBuildTools      exe @?= []
 
             _testMainIs       test @?= HsFilePath "quxTest/Main.hs" Standard
@@ -193,8 +198,13 @@ driverFunctionTest pkgIx srcDb comp = testGroup "createProject"
             _testLanguage     test @?= Haskell98
             _testOtherModules test @?= []
             _testOtherExts    test @?= []
-            _testDependencies test @?= []
+            _testDependencies test @?! []
             _testBuildTools   test @?= []
+
+            assertBool "The library should be a dependency of the executable" $
+              mkPackageNameDep (_optPkgName opts) `elem` _exeDependencies exe
+            assertBool "The library should be a dependency of the test executable" $
+              mkPackageNameDep (_optPkgName opts) `elem` _testDependencies test
 
           Right (ProjectSettings _ _ lib exe test, _) -> do
             lib  @?! Nothing
@@ -369,6 +379,11 @@ driverFunctionTest pkgIx srcDb comp = testGroup "createProject"
             _testDependencies test @?! []
             _testBuildTools   test @?= [mkStringyDep "happy:happy"]
 
+            assertBool "The library should be a dependency of the executable" $
+              mkPackageNameDep (_optPkgName opts) `elem` _exeDependencies exe
+            assertBool "The library should be a dependency of the test executable" $
+              mkPackageNameDep (_optPkgName opts) `elem` _testDependencies test
+
           Right (ProjectSettings _ _ lib exe test, _) -> do
             lib  @?! Nothing
             exe  @?! Nothing
@@ -499,6 +514,9 @@ driverFunctionTest pkgIx srcDb comp = testGroup "createProject"
             _testOtherExts    test @?= map EnableExtension [OverloadedStrings, LambdaCase, RankNTypes, RecordWildCards]
             _testDependencies test @?! []
             _testBuildTools   test @?= [mkStringyDep "happy:happy"]
+            
+            assertBool "The library should be a dependency of the test executable" $
+              mkPackageNameDep (_optPkgName opts) `elem` _testDependencies test
 
           Right (ProjectSettings _ _ lib exe test, _) -> do
             lib  @?! Nothing
@@ -639,6 +657,9 @@ driverFunctionTest pkgIx srcDb comp = testGroup "createProject"
             _exeOtherExts       exe @?= map EnableExtension [OverloadedStrings, LambdaCase, RankNTypes, RecordWildCards]
             _exeDependencies    exe @?! []
             _exeBuildTools      exe @?= [mkStringyDep "happy:happy"]
+
+            assertBool "The library should be a dependency of the executable" $
+              mkPackageNameDep (_optPkgName opts) `elem` _exeDependencies exe
 
           Right (ProjectSettings _ _ lib exe test, _) -> do
             lib  @?! Nothing

